hal/i2c_api.h

```diff
@@ -39,113 +41,227 @@
+void i2c_set_clock_stretching(i2c_t *obj, const bool enabled);
 
-/** Blocking reading data
+/** Configure the timeout duration in microseconds for blocking transmission
  *
- *  @param obj     The I2C object
- *  @param address 7-bit address (last bit is 1)
- *  @param data    The buffer for receiving
- *  @param length  Number of bytes to read
- *  @param stop    Stop to be generated after the transfer is done
- *  @return Number of read bytes
+ *  @param obj        The I2C object
+ *  @param timeout    Transmission timeout in microseconds.
+ *
+ *  @note If no timeout is set the default timeout is used.
+ *        Default timeout value is based on I2C frequency.
+ *        Byte timeout is computed as triple amount of time it would take
+ *        to send 10bit over I2C and is expressed by the formula:
+ *        byte_timeout = 3 * (1/frequency * 10 * 1000000)
  */
-int i2c_read(i2c_t *obj, int address, char *data, int length, int stop);
+void i2c_timeout(i2c_t *obj, uint32_t timeout);
+
+/** Send START command
+ *
+ *  @param obj The I2C object.
+ */
+void i2c_start(i2c_t *obj);
+
+/** Send STOP command
+ *
+ *  @param obj The I2C object
+ */
+void i2c_stop(i2c_t *obj);
 
 /** Blocking sending data
+ *
+ *  This function transmits data, when the peripheral is configured as Master to
+ *  the selected slave, and when configured as Slave transmits data to the
+ *  Master.
+ *
+ *  This function is blocking, it will return when the transfer is complete or a
+ *  timeout event is triggered. The number of bytes transmitted is returned by
+ *  the function after the operation is completed. Transmit operation cannot be
+ *  cancelled or aborted.
+ *
+ *  The data buffer must stay allocated during the duration of the transfer and
+ *  the contents must not be modified. The value of the specified `address` is
+ *  ignored when configured in slave mode, in master mode it contains the
+ *  address of the target peripheral. This is a 7-bit value unless 10-bit
+ *  addressing is configured and supported by the target.
+ *
+ *  When in master mode the operation consists of:
+ *   - Address the slave as a Master transmitter.
+ *   - Transmit data to the addressed slave.
+ *   - Generate a STOP condition if the specified `stop` field is true.
  *
  *  @param obj     The I2C object
- *  @param address 7-bit address (last bit is 0)
+ *  @param address 7/10-bit address (last bit is 0)
  *  @param data    The buffer for sending
  *  @param length  Number of bytes to write
- *  @param stop    Stop to be generated after the transfer is done
+ *  @param stop    If true, stop will be generated after the transfer is done
+ *
+ *  @note If the current platform supports multimaster operation the peripheral
+ *        will perform arbitration automatically when detects collision and
+ *        complete the transfer or return I2C_ERROR_ARBITRATION_LOST
+ *        when loses arbitration.
+ *
+ *        Additional time for arbitration or clock stretching should by count
```

Comment:
How much time?

Reply:
Actually we don't know how long it would take. It's up to user to decide how much time he can wait before failing